### PR TITLE
DDP Synchronization 

### PR DESCRIPTION
**Pull request by quinbez**
### Summary

Fixes DDP behavior for training and tuning by keeping all ranks synchronized after local pc weight updates, while preserving the existing two-phase tuning flow.

### Changes
**1. Manual PC weight synchronization in DDP** https://github.com/iCog-Labs-Dev/PC-Transformers/commit/49e9743fb45d81f6185a2da37ed2f0438bbfcfd5
Added manual synchronization after the local PC forward update so weights updated directly inside the pc path are averaged across ranks. This is needed because these updates bypass standard DDP gradient synchronization.

**2. DDP-safe lateral weight registration** https://github.com/iCog-Labs-Dev/PC-Transformers/commit/35660d487c56a4528855487b876578225dc619a2
Updated model initialization so lateral weights are registered before moving the model to device and before DDP wrapping. Also fixed the registered `layer_type` names to match the actual forward-pass names: `linear_attn`, `fc2`, and `linear_output`.

**3. DDP support for tuning** https://github.com/iCog-Labs-Dev/PC-Transformers/commit/c0544524e8ed53fefab31665ece6e07dfc03cf85
Added DDP support to `tuning/tune_model.py` while keeping Optuna controlled by rank 0. Rank 0 samples each trial, broadcasts the trial parameters to worker ranks, uses distributed loaders, and averages validation metrics across ranks. The two-phase tuning logic remains unchanged: Phase 1 minimizes energy, Phase 2 minimizes perplexity.

### Results

Training DDP sync check confirmed the expected behavior:
- Initial model state is synchronized across ranks: `max_param_diff_from_rank0=0.00000000e+00`
- Local PC update causes rank divergence: `max_param_diff_from_rank0=4.00000215e-02`
- Manual sync restores identical weights: `max_param_diff_from_rank0=0.00000000e+00`

The largest temporary divergence during training was in `embedding.word_embeddings.weight`, followed by output layer parameters, which is expected because each rank sees a different local batch.

Tuning DDP sync check also confirmed the expected behavior:
- Rank 0 started on `cuda:0`, rank 1 started on `cuda:1`
- Both ranks ran the same tuning trial with matching train/validation loader sizes
- Initial model state was synchronized: `max_param_diff_from_rank0=0.00000000e+00`
- Local PC update caused temporary divergence: `max_param_diff_from_rank0=1.03277713e-03`
- Manual sync restored identical weights: `max_param_diff_from_rank0=0.00000000e+00`

This confirms that both training and tuning now keep DDP ranks synchronized after local PC updates.

### CODE REVIEW

**Comment by Nardos24** [comment]
This branch has conflicts. Please resolve it first.

**Comment by quinbez** [comment]
> This branch has conflicts. Please resolve it first.

It has been resolved.

**Comment by Nardos24** [comment]
Since PyTorch's DDP is designed for backprop-based training and your PC updates bypass autograd entirely (no .backward() calls, no gradients to synchronize), the DDP wrapper in this PR is essentially idle; it contributes nothing to synchronization or learning. The sync_pc_weights function effectively replaces DDP's role by manually averaging weights using dist.all_reduce(). Given this, you could simplify the code by removing the DDP wrapper altogether and directly using dist for both communication and rank handling. And one thing to add isn't it possible to use the gradients from each step function (dE_dmu from each step function) before the weight update, rather than averaging the weights after the update?

**Comment by quinbez** [comment]
> Since PyTorch's DDP is designed for backprop-based training and your PC updates bypass autograd entirely (no .backward() calls, no gradients to synchronize), the DDP wrapper in this PR is essentially idle; it contributes nothing to synchronization or learning. The sync_pc_weights function effectively replaces DDP's role by manually averaging weights using dist.all_reduce(). Given this, you could simplify the code by removing the DDP wrapper altogether and directly using dist for both communication and rank handling. And one thing to add isn't it possible to use the gradients from each step function (dE_dmu from each step function) before the weight update, rather than averaging the weights after the update?

DDP has been replaced with explicit broadcast for distributed weight synchronization. `use_ddp` has also been renamed to `use_distributed` across all files to reflect that this is no longer tied to DDP specifically.

As for using the gradients from each step function, we could synchronize `update_w` before the weight update inside each step function, but in this architecture that is not sufficient on its own.

The main issue is that each `T` step also updates the latent state `x`: `x = x + inference_lr * error`. That updated `x` is then reused by other layers in the same step and future `T` steps. Since each rank processes different local data, their `x` values diverge immediately even if `update_w` is synchronized. Once `x` diverges, the next step produces different `dE_dmu` again, so the ranks drift apart anyway.

To make per-step gradient synchronization correct, we would also need to synchronize `x` after every layer and every `T` step, which is problematic for two reasons:
1. `x` is the latent state each rank is actively inferring from its own data shard. It is supposed to be different across ranks, that is the whole point of splitting the data. Syncing it would collapse all ranks into the same latent state, making distributed training no different from single GPU training.
2. Even if we accepted that, it would introduce roughly `2 × T × number_of_layers` all-reduce operations per batch, which becomes extremely expensive.

Additionally, `update_w` is passed through `Adam` in `PCOptimizer.step_param()`. `Adam` maintains per-rank optimizer state (`g1`, `g2`, `timestep`), so even identical synchronized updates can still produce different final weights unless the optimizer state is also synchronized. That is why averaging the final weights once after the full forward pass via `sync_pc_weights` is the most practical approach for this architecture.

File: utils/device_utils.py [hunk 1]
```diff
@@ -90,4 +91,29 @@ def setup_device():
...
+    with torch.no_grad():
+        for name, param in model.named_parameters():
+            if not torch.is_floating_point(param):
+                continue
+
+            dist.all_reduce(param.data, op=dist.ReduceOp.SUM)
+            param.data.div_(world_size)
+
+            if name.endswith("W_lateral"):
+                param.data.copy_(F.normalize(param.data, p=2, dim=1))
```
Comment (Nardos24) [inline utils/device_utils.py:119]: As we discussed previously, instead of averaging the weights by dividing by world_size, we can remove the param.data.div_(world_size) line and keep the sum from dist.all_reduce. This is mathematically equivalent to scaling the learning rate by the number of GPUs a common technique to make distributed training match sequential behavior. Removing this division keeps the code cleaner and avoids introducing an extra scaling factor that would otherwise need to be compensated elsewhere.  With this change, the only differences between distributed and sequential training are the speedup from parallelism and the natural distribution of data across GPUs not an artificial reduction in update magnitude. The performance impact of the update values themselves becomes zero, leaving only the intended benefits of distributed training.

Comment (quinbez) [reply utils/device_utils.py:119]: This has been addressed after further discussion. We agreed to keep the weight averaging approach as is to avoid weight explosions while preserving the current parallel learning behavior.

**Comment by Nardos24** [comment]
I have completed a comparison of training with and without DDP on our PC-Transformer model (6.20M parameters, 5 epochs, seed 42). Below is a summary of training speed, evaluation speed, perplexity, and energy for both configurations.

**Commands I used:**
- Single GPU: `python training.py --epochs=5 --seed=42` and `python eval.py`
- DDP (2 GPUs): `torchrun --nproc_per_node=2 training.py --epochs=5 --seed=42` and `torchrun --nproc_per_node=2 eval.py`


### **Table: Training and Evaluation Performance (Single GPU vs DDP 2 GPUs)**

| Metric | Single GPU | DDP (2 GPUs) |
|--------|-----------|--------------|
| **Training Time (5 epochs)** | 113.70 sec | 65.36 sec |
| **Evaluation Time** | 0.67 sec | 0.51 sec |
| **Batches per Epoch** | 145 | 73 |
| **Final Train Perplexity** | 943.49 | 2,863.39 |
| **Final Validation Perplexity** | 849.21 | 2,492.31 |
| **Final Train Energy** | 5,560,166.44 | 2,120,399.74 |
| **Final Validation Energy** | 5,507,574.02 | 2,160,559.58 |
| **Final Evaluation Perplexity** | 1,006.30 | 2,675.52 |
| **Final Evaluation Energy** | 5,445,119.37 | 2,347,572.67 |



So here are my observations

- **Batch processing and speed are improved** with DDP: training time reduced from 113.70 sec to 65.36 sec (1.74x faster), batches per epoch are properly distributed across GPUs, and evaluation time is slightly faster (0.67 sec → 0.51 sec).
- **However, this comes at a significant performance cost.** Final validation perplexity degrades from 849.21 (single GPU) to 2,492.31 (DDP), and final evaluation perplexity drops from 1,006.30 to 2,675.52. Energy values are lower.
- DDP synchronization works correctly (no crashes, no deadlocks), but the trade-off between speed and model quality is substantial for this architecture. Since our goal is to train at speed without degrading model performance (as we discussed), single-GPU training is currently the better choice for production and final model training. 

If your training results differ from mine, please let me know. Thanks.

**Comment by quinbez** [comment]
> I have completed a comparison of training with and without DDP on our PC-Transformer model (6.20M parameters, 5 epochs, seed 42). Below is a summary of training speed, evaluation speed, perplexity, and energy for both configurations.
> 
> **Commands I used:**
> 
> * Single GPU: `python training.py --epochs=5 --seed=42` and `python eval.py`
> * DDP (2 GPUs): `torchrun --nproc_per_node=2 training.py --epochs=5 --seed=42` and `torchrun --nproc_per_node=2 eval.py`
> 
> ### **Table: Training and Evaluation Performance (Single GPU vs DDP 2 GPUs)**
> Metric	Single GPU	DDP (2 GPUs)
> **Training Time (5 epochs)**	113.70 sec	65.36 sec
> **Evaluation Time**	0.67 sec	0.51 sec
> **Batches per Epoch**	145	73
> **Final Train Perplexity**	943.49	2,863.39
> **Final Validation Perplexity**	849.21	2,492.31
> **Final Train Energy**	5,560,166.44	2,120,399.74
> **Final Validation Energy**	5,507,574.02	2,160,559.58
> **Final Evaluation Perplexity**	1,006.30	2,675.52
> **Final Evaluation Energy**	5,445,119.37	2,347,572.67
> So here are my observations
> 
> * **Batch processing and speed are improved** with DDP: training time reduced from 113.70 sec to 65.36 sec (1.74x faster), batches per epoch are properly distributed across GPUs, and evaluation time is slightly faster (0.67 sec → 0.51 sec).
> * **However, this comes at a significant performance cost.** Final validation perplexity degrades from 849.21 (single GPU) to 2,492.31 (DDP), and final evaluation perplexity drops from 1,006.30 to 2,675.52. Energy values are lower.
> * DDP synchronization works correctly (no crashes, no deadlocks), but the trade-off between speed and model quality is substantial for this architecture. Since our goal is to train at speed without degrading model performance (as we discussed), single-GPU training is currently the better choice for production and final model training.
> 
> If your training results differ from mine, please let me know. Thanks.

Thank you for running the comparison. The performance difference is expected and is a known trade-off of distributed training, as we discussed. This PR specifically focuses on fixing the current synchronization issue when using multiple GPUs. Optimizing the trade-off between speed and model quality will be explored separately moving forward.

Also, we cannot conclusively say that single GPU training is better based on these results alone. Energy is lower on multi-GPU, while perplexity is lower on single GPU, so the metrics point in different directions. Further experimentation is needed before drawing conclusions about which configuration produces the better overall model, and that investigation will be part of future work.
